File: 048_Summary Ranges.cpp
/*
Summary Ranges

Given a sorted unique integer array nums, return the smallest sorted list of ranges
that cover all the numbers in the array exactly.
*/

#include <bits/stdc++.h>
using namespace std;

vector<string> summaryRanges(vector<int>& nums) {
    vector<string> result;
    if (nums.empty()) return result;
    
    int start = nums[0];
    int end = nums[0];
    
    for (int i = 1; i < (int)nums.size(); ++i) {
        if (nums[i] == end + 1) {
            end = nums[i];
        } else {
            if (start == end) {
                result.push_back(to_string(start));
            } else {
                result.push_back(to_string(start) + "->" + to_string(end));
            }
            start = end = nums[i];
        }
    }
    
    // Add the last range
    if (start == end) {
        result.push_back(to_string(start));
    } else {
        result.push_back(to_string(start) + "->" + to_string(end));
    }
    
    return result;
}

int main() {
    vector<int> n1 = {0, 1, 2, 4, 5, 7};
    auto r1 = summaryRanges(n1);
    cout << "Test 1: ";
    for (auto& s : r1) cout << s << " ";
    cout << endl; // ["0->2","4->5","7"]

    vector<int> n2 = {0, 2, 3, 4, 6, 8, 9};
    auto r2 = summaryRanges(n2);
    cout << "Test 2: ";
    for (auto& s : r2) cout << s << " ";
    cout << endl; // ["0","2->4","6","8->9"]
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1) excluding output

File: 048_Summary Ranges.py
"""
Summary Ranges

Given a sorted unique integer array nums, return the smallest sorted list of ranges
that cover all the numbers in the array exactly.
"""

def summaryRanges(nums: list[int]) -> list[str]:
    if not nums:
        return []
    
    result = []
    start = end = nums[0]
    
    for i in range(1, len(nums)):
        if nums[i] == end + 1:
            end = nums[i]
        else:
            if start == end:
                result.append(str(start))
            else:
                result.append(f"{start}->{end}")
            start = end = nums[i]
    
    # Add the last range
    if start == end:
        result.append(str(start))
    else:
        result.append(f"{start}->{end}")
    
    return result

if __name__ == "__main__":
    print("Test 1:", summaryRanges([0, 1, 2, 4, 5, 7]))  # ["0->2","4->5","7"]
    print("Test 2:", summaryRanges([0, 2, 3, 4, 6, 8, 9]))  # ["0","2->4","6","8->9"]

# Time Complexity: O(n)
# Space Complexity: O(1) excluding output
